src/import-utils.ts

```diff
@@ -0,0 +1,301 @@
+    return;
+  }
+
+  // Guard against inlining a import we've already processed.
+  if (reachedImports.has(importBundleUrl)) {
+    astUtils.removeElementAndNewline(htmlImport);
+    return;
+  }
+
+  // If the html import refers to a file which is bundled and has a different
+  // url, then lets just rewrite the href to point to the bundle url.
+  if (importBundleUrl !== docBundle.url) {
+    const relative =
+        urlUtils.relativeUrl(docUrl, importBundleUrl) || importBundleUrl;
+    dom5.setAttribute(htmlImport, 'href', relative);
+    reachedImports.add(importBundleUrl);
+    return;
+  }
+
+  const document =
+      dom5.nodeWalkAncestors(htmlImport, (node) => !node.parentNode)!;
+  const body = dom5.query(document, matchers.body)!;
+  const importSource = await loader(resolvedImportUrl).catch(err => {
+    throw new Error(`Unable to load ${resolvedImportUrl}: ${err.message}`);
+  });
+
+  // Is there a better way to get what we want other than using
+  // parseFragment?
+  const importDoc = parse5.parseFragment(importSource);
+  rewriteImportedUrls(basePath, importDoc, resolvedImportUrl, docUrl);
+  const nestedImports = dom5.queryAll(importDoc, matchers.htmlImport);
+
+  // Move all of the import doc content after the html import.
+  astUtils.insertAllBefore(
+      htmlImport.parentNode!, htmlImport, importDoc.childNodes!);
+  astUtils.removeElementAndNewline(htmlImport);
+
+  // If we've never seen this import before, lets add it to the set so we
+  // will deduplicate if we encounter it again.
+  reachedImports.add(resolvedImportUrl);
+
+  // Recursively process the nested imports.
+  for (const nestedImport of nestedImports) {
+    await inlineHtmlImport(
+        basePath,
+        docUrl,
+        nestedImport,
+        reachedImports,
+        docBundle,
+        manifest,
+        loader);
+  }
+}
+
+/**
+ * Inlines the contents of the document returned by the script tag's src url
+ * into the script tag content and removes the src attribute.
+ */
+export async function inlineScript(
+    docUrl: UrlString,
+    externalScript: ASTNode,
+    loader: (url: UrlString) => Promise<string>) {
+  const rawUrl: string = dom5.getAttribute(externalScript, 'src')!;
+  const resolvedUrl = urlLib.resolve(docUrl, rawUrl);
+  let script: string|undefined = undefined;
+  try {
+    script = await loader(resolvedUrl);
+  } catch (err) {
+    // If a script doesn't load, skip inlining.
+    // TODO(usergenic): Add plylog logging for load error.
+  }
+
+  if (script === undefined) {
+    return;
+  }
+
+  // Second argument 'true' tells encodeString to escape <script> tags.
+  const scriptContent = encodeString(script, true);
+  dom5.removeAttribute(externalScript, 'src');
+  dom5.setTextContent(externalScript, scriptContent);
+  return scriptContent;
+}
+
+/**
+ * Inlines the contents of the stylesheet returned by the link tag's href url
+ * into a style tag and removes the link tag.
+ */
+export async function inlineStylesheet(
```

Comment:
It might look like I'm passing around state, but I'm not really, or at least, I'm not passing around state of an object that's been yet been defined.  The passed arguments to these functions are specific to preparing a single bundled document and include the details necessary to add or do one thing to it as part of that process.  I could could name that state and build a class, but I don't have a great name for it yet.  I'll could add a TODO to the top of import-utils.ts to revisit and possibly refactor into a class called DocumentBuilder or something to help simplify the function calls, but ultimately these are functions that manipulate the dom by-way-of the elements they're handed and they don't change state of anything else, so I'm a little ambivalent.  I do agree that inlineHtmlImport has a truckload of arguments and I'd like to reduce that because eyesore.